File: apps/mobile/src/models/Transaction.ts
import { Model } from '@nozbe/watermelondb';
import { field, date, readonly, relation } from '@nozbe/watermelondb/decorators';
import type { Associations } from '@nozbe/watermelondb/Model';
import type Family from './Family';
import type Wallet from './Wallet';

export type TransactionType = 'credit' | 'debit';
export type TransactionSourceType = 'allowance' | 'request' | 'manual' | 'adjustment';

export default class Transaction extends Model {
  static table = 'transactions';

  static associations: Associations = {
    families: { type: 'belongs_to', key: 'family_id' },
    wallets: { type: 'belongs_to', key: 'wallet_id' },
  };

  @field('family_id') familyId!: string;
  @field('wallet_id') walletId!: string;
  @field('type') type!: TransactionType;
  @field('amount_cents') amountCents!: number;
  @field('source_type') sourceType!: TransactionSourceType;
  @field('source_id') sourceId?: number;
  @field('category') category?: string;
  @date('occurred_at') occurredAt!: Date;
  @field('notes') notes?: string;

  @readonly @date('created_at') createdAt!: Date;

  @relation('families', 'family_id') family!: Family;
  @relation('wallets', 'wallet_id') wallet!: Wallet;
}

File: src-tauri/src/models/transaction.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum TransactionType {
    Credit,
    Debit,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum SourceType {
    Allowance,
    Request,
    Manual,
    Adjustment,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct Transaction {
    pub id: i64,
    pub uuid: String,
    pub family_id: i64,
    pub wallet_id: i64,
    #[serde(rename = "type")]
    #[sqlx(rename = "type")]
    pub type_: String,
    pub amount_cents: i64,
    pub source_type: String,
    pub source_id: Option<i64>,
    pub category: Option<String>,
    pub occurred_at: String,
    pub notes: Option<String>,
    pub sync_version: i64,
    pub last_synced_at: Option<String>,
    pub is_deleted: i64,
    pub created_at: String,
}
